File: agent/skills/rain-issue-consumer-card/examples/issue-card-end-to-end.go
// Issue a Rain consumer card end-to-end (Go).
//
//	1. create application  2. await KYC  3. create virtual card
//	4. retrieve + decrypt secrets
//
// Run (sandbox):
//
//	export RAIN_API_KEY=<sandbox key>
//	go run issue-card-end-to-end.go
//
// Uses the sandbox `approved`-last-name shortcut. Prints only last4.
//
// The session-id + decrypt helpers live in ../scripts (package cardcrypto). In a real
// project, import that package; here the calls are shown inline as cardcrypto.*.
package main

import (
	"context"
	"errors"
	"fmt"
	"os"
	"time"

	rainsdk "github.com/SignifyHQ/rain-sdk-go"
	"github.com/SignifyHQ/rain-sdk-go/option"

	cardcrypto "example.com/rain/scripts" // the ../scripts package
)

var terminalFail = map[string]bool{"denied": true, "locked": true, "canceled": true, "exempt": true}
var actionRequired = map[string]bool{"needsVerification": true, "needsInformation": true}

func awaitApproval(ctx context.Context, client *rainsdk.Client, userID string) error {
	deadline := time.Now().Add(120 * time.Second)
	for time.Now().Before(deadline) {
		app, err := client.Applications.User.Get(ctx, userID)
		if err != nil {
			return err
		}
		switch status := string(app.ApplicationStatus); {
		case status == "approved":
			return nil
		case terminalFail[status]:
			return fmt.Errorf("terminal status: %s", status)
		case actionRequired[status]:
			return fmt.Errorf("action required (%s): redirect to applicationCompletionLink", status)
		}
		time.Sleep(4 * time.Second) // pending | manualReview | notStarted
	}
	return errors.New("timed out waiting for KYC")
}

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	defer cancel()

	client := rainsdk.NewClient(
		option.WithAPIKey(os.Getenv("RAIN_API_KEY")),
		option.WithEnvironmentDev(), // option.WithEnvironmentProduction() for live
	)

	// 1. create application (sandbox: last name contains "approved")
	app, err := client.Applications.User.New(ctx, rainsdk.ApplicationUserNewParams{
		IPAddress:                rainsdk.String("203.0.113.10"),
		Occupation:               rainsdk.String("15-1252"),
		AnnualSalary:             rainsdk.String("50000-100000"),
		AccountPurpose:           rainsdk.String("web3Payments"),
		ExpectedMonthlyVolume:    rainsdk.String("1000-5000"),
		IsTermsOfServiceAccepted: rainsdk.Bool(true),
		WalletAddress:            rainsdk.String("0x1234567890abcdef1234567890abcdef12345678"),
		FirstName:                rainsdk.String("Jane"),
		LastName:                 rainsdk.String("Doe approved"),
		BirthDate:                rainsdk.String("1990-04-15"),
		NationalID:               rainsdk.String("123456789"),
		CountryOfIssue:           rainsdk.String("US"),
		Email:                    rainsdk.String(fmt.Sprintf("jane.doe.%d@example.com", time.Now().Unix())),
	}, option.WithHeader("Idempotency-Key", randomKey()))
	if err != nil {
		panic(err)
	}
	userID := app.ID
	fmt.Println("application:", userID, "status:", app.ApplicationStatus)

	// 2. await KYC
	if err := awaitApproval(ctx, client, userID); err != nil {
		panic(err)
	}
	fmt.Println("KYC approved")

	// 3. create a virtual card ($500 / rolling 30 days)
	card, err := client.Users.NewCard(ctx, userID, rainsdk.UserNewCardParams{
		Type: rainsdk.UserNewCardParamsTypeVirtual,
		Limit: rainsdk.IssuingCardLimitParam{
			Amount:    rainsdk.Int(50000),
			Frequency: "per30DayPeriod",
		},
	}, option.WithHeader("Idempotency-Key", randomKey()))
	if err != nil {
		panic(err)
	}
	cardID := card.ID
	fmt.Println("card created:", cardID)

	// 4. retrieve + decrypt secrets (SessionId key + AES-128-GCM)
	secretKey, sessionID, err := cardcrypto.GenerateSessionID(cardcrypto.DevSessionIDPublicKey, "")
	if err != nil {
		panic(err)
	}
	secrets, err := client.Cards.GetSecrets(ctx, cardID, option.WithHeader("SessionId", sessionID))
	if err != nil {
		panic(err)
	}
	pan, err := cardcrypto.DecryptSecret(secrets.EncryptedPan.Data, secrets.EncryptedPan.IV, secretKey)
	if err != nil {
		panic(err)
	}
	_, _ = cardcrypto.DecryptSecret(secrets.EncryptedCvc.Data, secrets.EncryptedCvc.IV, secretKey)
	fmt.Println("issued card ending", pan[len(pan)-4:]) // never log the full PAN/CVC
}

func randomKey() string {
	return fmt.Sprintf("%d-%d", time.Now().UnixNano(), os.Getpid())
}

File: agent/skills/rain-issue-consumer-card/examples/issue-card-end-to-end.ts
/**
 * Issue a Rain consumer card end-to-end (TypeScript).
 *
 *   1. create application  2. upload docs  3. await KYC
 *   4. create virtual card 5. retrieve + decrypt secrets
 *
 * Run (sandbox):
 *   export RAIN_API_KEY=<sandbox key>
 *   npx tsx issue-card-end-to-end.ts
 *
 * Uses the sandbox `approved`-last-name shortcut so KYC auto-approves.
 * Prints only last4 — never the full PAN/CVC.
 */
import crypto from "node:crypto";
import { createReadStream } from "node:fs";
import Rain from "@rainapi/rain-sdk";

import { generateSessionId, DEV_SESSIONID_PUBLIC_KEY } from "../scripts/generate-session-id";
import { decryptSecret } from "../scripts/decrypt-card-secret";

const client = new Rain({
  apiKey: process.env["RAIN_API_KEY"],
  environment: "dev", // 'production' for live
});

const TERMINAL_FAIL = ["denied", "locked", "canceled", "exempt"];
const ACTION_REQUIRED = ["needsVerification", "needsInformation"];

async function awaitApproval(userId: string, timeoutMs = 120_000, intervalMs = 4_000) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const app = await client.applications.user.get(userId);
    const status = app.applicationStatus;
    if (status === "approved") return app;
    if (TERMINAL_FAIL.includes(status)) throw new Error(`terminal status: ${status}`);
    if (ACTION_REQUIRED.includes(status)) {
      // send the user to app.applicationCompletionLink (pass all params + a redirect)
      throw new Error(`action required (${status}): redirect user to applicationCompletionLink`);
    }
    await new Promise((r) => setTimeout(r, intervalMs)); // pending | manualReview | notStarted
  }
  throw new Error("timed out waiting for KYC");
}

async function main() {
  // 1. Create application (sandbox: lastName contains "approved" -> auto-approve)
  const application = await client.applications.user.create(
    {
      ipAddress: "203.0.113.10",
      occupation: "15-1252",
      annualSalary: "50000-100000",
      accountPurpose: "web3Payments",
      expectedMonthlyVolume: "1000-5000",
      isTermsOfServiceAccepted: true,
      // Rain-Managed only: one wallet field required
      walletAddress: "0x1234567890abcdef1234567890abcdef12345678",
      firstName: "Jane",
      lastName: "Doe approved",
      birthDate: "1990-04-15",
      nationalId: "123456789",
      countryOfIssue: "US",
      email: `jane.doe.${Date.now()}@example.com`,
      address: { line1: "123 Main St", city: "New York", region: "NY", postalCode: "10001", countryCode: "US" },
    },
    { headers: { "Idempotency-Key": crypto.randomUUID() } },
  );
  const userId = application.id;
  console.log("application created:", userId, "status:", application.applicationStatus);

  // 2. Upload documents (skip if already approved via the sandbox shortcut)
  if (application.applicationStatus !== "approved" && process.env["UPLOAD_DOCS"]) {
    await client.applications.user.uploadDocument(userId, {
      document: createReadStream("./passport.png"),
      type: "passport",
      countryCode: "US",
    });
    await client.applications.user.uploadDocument(userId, {
      document: createReadStream("./selfie.jpg"),
      type: "selfie",
    });
  }

  // 3. Await KYC
  await awaitApproval(userId);
  console.log("KYC approved");

  // 4. Create a virtual card ($500 / rolling 30 days)
  const card = await client.users.createCard(
    userId,
    {
      type: "virtual",
      limit: { amount: 50_000, frequency: "per30DayPeriod" },
      configuration: { displayName: "JANE DOE" },
    },
    { headers: { "Idempotency-Key": crypto.randomUUID() } },
  );
  const cardId = card.id;
  console.log("card created:", cardId);

  // 5. Retrieve + decrypt secrets (SessionId key + AES-128-GCM)
  const { sessionId, secretKey } = generateSessionId(DEV_SESSIONID_PUBLIC_KEY);
  const secrets = await client.cards.getSecrets(cardId, { headers: { SessionId: sessionId } });
  const pan = decryptSecret(secrets.encryptedPan.data, secrets.encryptedPan.iv, secretKey);
  const cvc = decryptSecret(secrets.encryptedCvc.data, secrets.encryptedCvc.iv, secretKey);
  void cvc; // hand to your secure surface; never log it
  console.log("issued card ending", pan.slice(-4)); // never log the full PAN
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
